File: frontend/src/types/SearxResult.ts
import JsonApi from "@/types/JsonApi";
import Source from "@/types/Source";

export interface SearxResultAttributes {
  title: string;
  url: string;
  engine: string;
  parsedUrl: string[]; // TODO: this is actually a tuple (protocol, domain, subpath, <something>, query string, <something>)
  engines: string[];
  positions: number[];
  content: string;
  isOnion: boolean;
  score: number;
  category: string;
  prettyUrl: string;
  sanitizedPath: string;
  pathSegments: string[];
  pathFamily: string[];
}

export type SearxResultRelationships = JsonApi.Relationships<{
  sources: JsonApi.CollectionRelationshipDetails;
}>;

export type SearxResultLinks = undefined;
export type SearxResultMeta = undefined;

type SearxResult = JsonApi.ResourceData<
  "searxResult",
  SearxResultAttributes,
  SearxResultRelationships,
  SearxResultLinks,
  SearxResultMeta
>;

export default SearxResult;

export type SearxResultItemResponse<
  M extends JsonApi.Meta | undefined = undefined,
> = JsonApi.ItemResponse<
  SearxResult,
  Source[],
  undefined,
  M
>;

export type SearxResultCollectionResponse<
  M extends JsonApi.Meta | undefined = undefined,
> = JsonApi.CollectionResponse<
  SearxResult[],
  Source[],
  undefined, // TODO: This should have pagination links
  M // TODO: This should have infoboxes and result count and stuff
>;

File: backend/app/models/searx_result.rb
#
# NOTE: This model class is a PORO, not an ActiveRecord model class.
#
class SearxResult
  SEARX_RESULT_PROPERTIES = %i[
    category
    content
    engine
    engines
    is_onion
    parsed_url
    positions
    pretty_url
    score
    title
    url
  ].freeze

  attr_accessor(*SEARX_RESULT_PROPERTIES)

  attr_accessor(*%i[
    source_ids
  ])

  def initialize(raw_attrs = {})
    attrs = raw_attrs.symbolize_keys
    whitelisted_attrs = attrs.slice(*SEARX_RESULT_PROPERTIES)
    whitelisted_attrs.each { |key, value| send("#{key}=", value) }
    self.source_ids = attrs[:source_ids] || []
  end

  def id
    @id ||= Digest::MD5.hexdigest(url)
  end

  def sanitized_path
    Source.sanitize_path(url)
  end

  def path_segments
    Source.path_segments(sanitized_path)
  end

  def path_family
    Source.path_family(sanitized_path)
  end

  # def sources
  #   @sources ||= Source.source_family(sanitized_path)
  # end

  # def source_ids
  #   sources.ids
  # end
end
